Unclear query response to debug JSON_MATCH on one column without json index

Hi, the Pinot Controller query response is unclear to show the root cause when we use JSON_MATCH on one column without proper json index. However, we can find more information from server logs. It is good to have a clear response message from Pinot Controller for debugging.

This is the query response from Pinot Controller:
`[
  {
    "errorCode": 200,
    "message": "QueryExecutionError:\njava.lang.RuntimeException: Caught exception while running CombinePlanNode.\n\tat org.apache.pinot.core.plan.CombinePlanNode.run(CombinePlanNode.java:157)\n\tat org.apache.pinot.core.plan.InstanceResponsePlanNode.run(InstanceResponsePlanNode.java:33)\n\tat org.apache.pinot.core.plan.GlobalPlanImplV0.execute(GlobalPlanImplV0.java:45)\n\tat org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:296)\n\tat org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:216)\n\tat org.apache.pinot.core.query.executor.QueryExecutor.processQuery(QueryExecutor.java:60)\n\tat org.apache.pinot.core.query.scheduler.QueryScheduler.processQueryAndSerialize(QueryScheduler.java:155)\n\tat org.apache.pinot.core.query.scheduler.QueryScheduler.lambda$createQueryFutureTask$0(QueryScheduler.java:139)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)\n\tat com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)\n\tat com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)"
  },
  {
    "errorCode": 200,
    "message": "QueryExecutionError:\njava.lang.RuntimeException: Caught exception while running CombinePlanNode.\n\tat org.apache.pinot.core.plan.CombinePlanNode.run(CombinePlanNode.java:157)\n\tat org.apache.pinot.core.plan.InstanceResponsePlanNode.run(InstanceResponsePlanNode.java:33)\n\tat org.apache.pinot.core.plan.GlobalPlanImplV0.execute(GlobalPlanImplV0.java:45)\n\tat org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:296)\n\tat org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:216)\n\tat org.apache.pinot.core.query.executor.QueryExecutor.processQuery(QueryExecutor.java:60)\n\tat org.apache.pinot.core.query.scheduler.QueryScheduler.processQueryAndSerialize(QueryScheduler.java:155)\n\tat org.apache.pinot.core.query.scheduler.QueryScheduler.lambda$createQueryFutureTask$0(QueryScheduler.java:139)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)\n\tat com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)\n\tat com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)"
  },
  {
    "errorCode": 200,
    "message": "QueryExecutionError:\njava.lang.RuntimeException: Caught exception while running CombinePlanNode.\n\tat org.apache.pinot.core.plan.CombinePlanNode.run(CombinePlanNode.java:157)\n\tat org.apache.pinot.core.plan.InstanceResponsePlanNode.run(InstanceResponsePlanNode.java:33)\n\tat org.apache.pinot.core.plan.GlobalPlanImplV0.execute(GlobalPlanImplV0.java:45)\n\tat org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:296)\n\tat org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:216)\n\tat org.apache.pinot.core.query.executor.QueryExecutor.processQuery(QueryExecutor.java:60)\n\tat org.apache.pinot.core.query.scheduler.QueryScheduler.processQueryAndSerialize(QueryScheduler.java:155)\n\tat org.apache.pinot.core.query.scheduler.QueryScheduler.lambda$createQueryFutureTask$0(QueryScheduler.java:139)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)\n\tat com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)\n\tat com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)"
  }
]`

The root cause can be found from server logs which is more clear as below:
`2021-09-17 21:11:56.103 [pqr-29] ERROR o.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl  - Exception processing requestId 9210057
java.lang.RuntimeException: Caught exception while running CombinePlanNode.
        at org.apache.pinot.core.plan.CombinePlanNode.run(CombinePlanNode.java:157)
        at org.apache.pinot.core.plan.InstanceResponsePlanNode.run(InstanceResponsePlanNode.java:33)
        at org.apache.pinot.core.plan.GlobalPlanImplV0.execute(GlobalPlanImplV0.java:45)
        at org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:296)
        at org.apache.pinot.core.query.executor.ServerQueryExecutorV1Impl.processQuery(ServerQueryExecutorV1Impl.java:216)
        at org.apache.pinot.core.query.executor.QueryExecutor.processQuery(QueryExecutor.java:60)
        at org.apache.pinot.core.query.scheduler.QueryScheduler.processQueryAndSerialize(QueryScheduler.java:155)
        at org.apache.pinot.core.query.scheduler.QueryScheduler.lambda$createQueryFutureTask$0(QueryScheduler.java:139)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Cannot apply JSON_MATCH on column: metadata without json index
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:206)
        at org.apache.pinot.core.plan.CombinePlanNode.run(CombinePlanNode.java:147)
        ... 15 common frames omitted
Caused by: java.lang.IllegalStateException: Cannot apply JSON_MATCH on column: metadata without json index
        at com.google.common.base.Preconditions.checkState(Preconditions.java:585)
        at org.apache.pinot.core.plan.FilterPlanNode.constructPhysicalOperator(FilterPlanNode.java:206)
        at org.apache.pinot.core.plan.FilterPlanNode.run(FilterPlanNode.java:80)
        at org.apache.pinot.core.plan.DocIdSetPlanNode.run(DocIdSetPlanNode.java:41)
        at org.apache.pinot.core.plan.ProjectionPlanNode.run(ProjectionPlanNode.java:52)
        at org.apache.pinot.core.plan.TransformPlanNode.run(TransformPlanNode.java:52)
        at org.apache.pinot.core.plan.SelectionPlanNode.run(SelectionPlanNode.java:83)
        at org.apache.pinot.core.plan.CombinePlanNode$1.callJob(CombinePlanNode.java:133)
        at org.apache.pinot.core.plan.CombinePlanNode$1.callJob(CombinePlanNode.java:119)
        at org.apache.pinot.core.util.trace.TraceCallable.call(TraceCallable.java:44)
        ... 8 common frames omitted`

--- a/pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java
+++ b/pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java
@@ -20,6 +20,7 @@ package org.apache.pinot.common.exception;
 
 import java.io.PrintWriter;
 import java.io.StringWriter;
+import java.util.regex.Pattern;
 import org.apache.pinot.common.response.ProcessingException;
 
 
@@ -27,7 +28,15 @@ public class QueryException {
   private QueryException() {
   }
 
-  private static int _maxLinesOfStackTrace = 15;
+  private static int _maxLinesOfStackTrace = 5;
+
+  /**
+   * The 2 regexp below must conform with JDK's internal implementation in {@link Throwable#printStackTrace()}.
+   */
+  private static final Pattern CAUSE_CAPTION_REGEXP = Pattern.compile("^([\\t]*)Caused by: ");
+  private static final Pattern SUPPRESSED_CAPTION_REGEXP = Pattern.compile("^([\\t]*)Suppressed: ");
+
+  private static final String OMITTED_SIGNAL = "...";
 
   // TODO: config max lines of stack trace if necessary. The config should be on instance level.
   public static void setMaxLinesOfStackTrace(int maxLinesOfStackTrace) {
@@ -144,17 +153,29 @@ public class QueryException {
     return copiedProcessingException;
   }
 
-  public static String getTruncatedStackTrace(Exception exception) {
+  public static String getTruncatedStackTrace(Throwable exception) {
     StringWriter stringWriter = new StringWriter();
     exception.printStackTrace(new PrintWriter(stringWriter));
     String fullStackTrace = stringWriter.toString();
     String[] lines = fullStackTrace.split("\n");
-    int numLinesOfStackTrace = Math.min(lines.length, _maxLinesOfStackTrace);
-    int lengthOfStackTrace = numLinesOfStackTrace - 1;
-    for (int i = 0; i < numLinesOfStackTrace; i++) {
-      lengthOfStackTrace += lines[i].length();
+    // exception should at least have one line, no need to check here.
+    StringBuilder sb = new StringBuilder(lines[0]);
+    int lengthOfStackTrace = 1;
+    for (int i = 1; i < lines.length; i++) {
+      if (CAUSE_CAPTION_REGEXP.matcher(lines[i]).find() || SUPPRESSED_CAPTION_REGEXP.matcher(lines[i]).find()) {
+        // reset stack trace print counter when a new cause or suppressed Throwable were found.
+        if (lengthOfStackTrace >= _maxLinesOfStackTrace) {
+          sb.append("\n").append(OMITTED_SIGNAL);
+        }
+        sb.append("\n").append(lines[i]);
+        lengthOfStackTrace = 1;
+      } else if (lengthOfStackTrace < _maxLinesOfStackTrace) {
+        // only print numLinesOfStackTrace stack trace and ignore any additional lines.
+        sb.append("\n").append(lines[i]);
+        lengthOfStackTrace++;
+      }
     }
-    return fullStackTrace.substring(0, lengthOfStackTrace);
+    return sb.toString();
   }
 
   /**

--- a/pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java
+++ b/pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java
@@ -21,6 +21,7 @@ package org.apache.pinot.common.exception;
 import java.io.PrintWriter;
 import java.io.StringWriter;
 import java.util.regex.Pattern;
+import org.apache.commons.lang3.StringUtils;
 import org.apache.pinot.common.response.ProcessingException;
 
 
@@ -28,7 +29,7 @@ public class QueryException {
   private QueryException() {
   }
 
-  private static int _maxLinesOfStackTrace = 5;
+  private static int _maxLinesOfStackTracePerFrame = 5;
 
   /**
    * The 2 regexp below must conform with JDK's internal implementation in {@link Throwable#printStackTrace()}.
@@ -39,8 +40,8 @@ public class QueryException {
   private static final String OMITTED_SIGNAL = "...";
 
   // TODO: config max lines of stack trace if necessary. The config should be on instance level.
-  public static void setMaxLinesOfStackTrace(int maxLinesOfStackTrace) {
-    _maxLinesOfStackTrace = maxLinesOfStackTrace;
+  public static void setMaxLinesOfStackTrace(int maxLinesOfStackTracePerFrame) {
+    _maxLinesOfStackTracePerFrame = maxLinesOfStackTracePerFrame;
   }
 
   // TODO: several ProcessingExceptions are never used, clean them up.
@@ -157,22 +158,22 @@ public class QueryException {
     StringWriter stringWriter = new StringWriter();
     exception.printStackTrace(new PrintWriter(stringWriter));
     String fullStackTrace = stringWriter.toString();
-    String[] lines = fullStackTrace.split("\n");
+    String[] lines = StringUtils.split(fullStackTrace, '\n');
     // exception should at least have one line, no need to check here.
     StringBuilder sb = new StringBuilder(lines[0]);
-    int lengthOfStackTrace = 1;
+    int lengthOfStackTracePerFrame = 1;
     for (int i = 1; i < lines.length; i++) {
       if (CAUSE_CAPTION_REGEXP.matcher(lines[i]).find() || SUPPRESSED_CAPTION_REGEXP.matcher(lines[i]).find()) {
         // reset stack trace print counter when a new cause or suppressed Throwable were found.
-        if (lengthOfStackTrace >= _maxLinesOfStackTrace) {
-          sb.append("\n").append(OMITTED_SIGNAL);
+        if (lengthOfStackTracePerFrame >= _maxLinesOfStackTracePerFrame) {
+          sb.append('\n').append(OMITTED_SIGNAL);
         }
-        sb.append("\n").append(lines[i]);
-        lengthOfStackTrace = 1;
-      } else if (lengthOfStackTrace < _maxLinesOfStackTrace) {
+        sb.append('\n').append(lines[i]);
+        lengthOfStackTracePerFrame = 1;
+      } else if (lengthOfStackTracePerFrame < _maxLinesOfStackTracePerFrame) {
         // only print numLinesOfStackTrace stack trace and ignore any additional lines.
-        sb.append("\n").append(lines[i]);
-        lengthOfStackTrace++;
+        sb.append('\n').append(lines[i]);
+        lengthOfStackTracePerFrame++;
       }
     }
     return sb.toString();

--- a/pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java
+++ b/pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java
@@ -161,19 +161,19 @@ public class QueryException {
     String[] lines = StringUtils.split(fullStackTrace, '\n');
     // exception should at least have one line, no need to check here.
     StringBuilder sb = new StringBuilder(lines[0]);
-    int lengthOfStackTracePerFrame = 1;
+    int lineOfStackTracePerFrame = 1;
     for (int i = 1; i < lines.length; i++) {
       if (CAUSE_CAPTION_REGEXP.matcher(lines[i]).find() || SUPPRESSED_CAPTION_REGEXP.matcher(lines[i]).find()) {
         // reset stack trace print counter when a new cause or suppressed Throwable were found.
-        if (lengthOfStackTracePerFrame >= _maxLinesOfStackTracePerFrame) {
+        if (lineOfStackTracePerFrame >= _maxLinesOfStackTracePerFrame) {
           sb.append('\n').append(OMITTED_SIGNAL);
         }
         sb.append('\n').append(lines[i]);
-        lengthOfStackTracePerFrame = 1;
-      } else if (lengthOfStackTracePerFrame < _maxLinesOfStackTracePerFrame) {
+        lineOfStackTracePerFrame = 1;
+      } else if (lineOfStackTracePerFrame < _maxLinesOfStackTracePerFrame) {
         // only print numLinesOfStackTrace stack trace and ignore any additional lines.
         sb.append('\n').append(lines[i]);
-        lengthOfStackTracePerFrame++;
+        lineOfStackTracePerFrame++;
       }
     }
     return sb.toString();

--- a/pinot-common/src/test/java/org/apache/pinot/common/exception/QueryExceptionTest.java
+++ b/pinot-common/src/test/java/org/apache/pinot/common/exception/QueryExceptionTest.java
@@ -9,24 +9,27 @@ import static org.testng.AssertJUnit.assertEquals;
 public class QueryExceptionTest {
 
   @Test
-  public void testExceptionMessage() throws Exception {
+  public void testExceptionMessage() {
     Exception exception = new UnsupportedOperationException("Caught exception.");
     ProcessingException processingException =
         QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, exception);
-    // should match QueryException._maxLinesOfStackTrace + 1 lines for the top level wrapper.
+
+    // there's one more 1 lines for the top level wrapper QUERY_EXECUTION_ERROR
     assertEquals(5 + 1, processingException.getMessage().split("\n").length);
 
     Exception withSuppressedException = new IllegalStateException("Suppressed exception");
     withSuppressedException.addSuppressed(processingException);
     ProcessingException withSuppressedProcessingException =
         QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, withSuppressedException);
-    // should match QueryException._maxLinesOfStackTrace * 2 + 1 line separator + 1 lines for the top level wrapper.
+
+    // QueryException._maxLinesOfStackTracePerFrame * 2 + 1 line separator + 1 QUERY_EXECUTION_ERROR wrapper.
     assertEquals(5 * 2 + 1 + 1, withSuppressedProcessingException.getMessage().split("\n").length);
 
     Exception withNestedException = new IllegalStateException("Outer exception", withSuppressedProcessingException);
     ProcessingException withNestedProcessingException =
         QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, withNestedException);
-    // should match QueryException._maxLinesOfStackTrace * 3 + 2 line separator + 1 lines for the top level wrapper.
+
+    // QueryException._maxLinesOfStackTracePerFrame * 3 + 2 line separators + 1 QUERY_EXECUTION_ERROR wrapper.
     assertEquals(5 * 3 + 2 + 1, withNestedProcessingException.getMessage().split("\n").length);
   }
 }

--- a/pinot-common/src/test/java/org/apache/pinot/common/exception/QueryExceptionTest.java
+++ b/pinot-common/src/test/java/org/apache/pinot/common/exception/QueryExceptionTest.java
@@ -1,0 +1,32 @@
+package org.apache.pinot.common.exception;
+
+import org.apache.pinot.common.response.ProcessingException;
+import org.testng.annotations.Test;
+
+import static org.testng.AssertJUnit.assertEquals;
+
+
+public class QueryExceptionTest {
+
+  @Test
+  public void testExceptionMessage() throws Exception {
+    Exception exception = new UnsupportedOperationException("Caught exception.");
+    ProcessingException processingException =
+        QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, exception);
+    // should match QueryException._maxLinesOfStackTrace + 1 lines for the top level wrapper.
+    assertEquals(5 + 1, processingException.getMessage().split("\n").length);
+
+    Exception withSuppressedException = new IllegalStateException("Suppressed exception");
+    withSuppressedException.addSuppressed(processingException);
+    ProcessingException withSuppressedProcessingException =
+        QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, withSuppressedException);
+    // should match QueryException._maxLinesOfStackTrace * 2 + 1 line separator + 1 lines for the top level wrapper.
+    assertEquals(5 * 2 + 1 + 1, withSuppressedProcessingException.getMessage().split("\n").length);
+
+    Exception withNestedException = new IllegalStateException("Outer exception", withSuppressedProcessingException);
+    ProcessingException withNestedProcessingException =
+        QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, withNestedException);
+    // should match QueryException._maxLinesOfStackTrace * 3 + 2 line separator + 1 lines for the top level wrapper.
+    assertEquals(5 * 3 + 2 + 1, withNestedProcessingException.getMessage().split("\n").length);
+  }
+}

--- a/pinot-common/src/test/java/org/apache/pinot/common/exception/QueryExceptionTest.java
+++ b/pinot-common/src/test/java/org/apache/pinot/common/exception/QueryExceptionTest.java
@@ -1,3 +1,21 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
 package org.apache.pinot.common.exception;
 
 import org.apache.pinot.common.response.ProcessingException;

